SU2_CFD/src/output/filewriter/CCGNSFileWriter.cpp

```diff
@@ -0,0 +1,234 @@
+    recvBufferField.resize(recvSize);
+
+    SU2_MPI::Recv(recvBufferField.data(), recvSize * sizeof(dataPrecision), MPI_CHAR, i, 0, SU2_MPI::GetComm(),
+                  MPI_STATUS_IGNORE);
+    if (recvSize <= 0) continue;
+    if (isCoord) {
+      int CoordinateNumber;
+      CallCGNS(cg_coord_partial_write(cgnsFileID, cgnsBase, cgnsZone, dataType, FieldName.c_str(), &nodeBegin, &nodeEnd,
+                                      recvBufferField.data(), &CoordinateNumber));
+    } else {
+      int fieldNumber;
+      CallCGNS(cg_field_partial_write(cgnsFileID, cgnsBase, cgnsZone, cgnsFields, dataType, FieldName.c_str(),
+                                      &nodeBegin, &nodeEnd, recvBufferField.data(), &fieldNumber));
+    }
+  }
+}
+
+void CCGNSFileWriter::WriteConnectivity(GEO_TYPE type, const string& SectionName) {
+  const auto nTotElem = dataSorter->GetnElemGlobal(type);
+  if (nTotElem == 0) return;
+
+  /*--- Create a new CGNS node to store connectivity. ---*/
+  const auto elementType = GetCGNSType(type);
+
+  cgsize_t firstElem = cumulative + 1;
+  cgsize_t endElem = cumulative + static_cast<cgsize_t>(nTotElem);
+
+  int cgnsSection;
+  if (rank == MASTER_NODE)
+    CallCGNS(cg_section_partial_write(cgnsFileID, cgnsBase, cgnsZone, SectionName.c_str(), elementType, firstElem,
+                                      endElem, 0, &cgnsSection));
+
+  /*--- Retrieve element distribution among processes. ---*/
+  const auto nLocalElem = dataSorter->GetnElem(type);
+
+  vector<unsigned long> distElem(size);
+
+  SU2_MPI::Allgather(&nLocalElem, 1, MPI_UNSIGNED_LONG, distElem.data(), 1, MPI_UNSIGNED_LONG, SU2_MPI::GetComm());
+
+  firstElem = cumulative + 1;
+  endElem = cumulative + static_cast<cgsize_t>(distElem[rank]);
+
+  /*--- Connectivity is stored in send buffer. ---*/
+  const auto nPointsElem = nPointsOfElementType(type);
+  sendBufferConnectivity.resize(nLocalElem * nPointsElem);
+
+  for (unsigned long iElem = 0; iElem < nLocalElem; iElem++) {
+    for (unsigned long iPoint = 0; iPoint < nPointsElem; iPoint++) {
+      sendBufferConnectivity[iPoint + nPointsElem * iElem] =
+          static_cast<cgsize_t>(dataSorter->GetElem_Connectivity(type, iElem, iPoint));
+    }
+  }
+
+  const auto bufferSize = static_cast<int>(nLocalElem * nPointsElem * sizeof(cgsize_t));
+  if (rank != MASTER_NODE) {
+    SU2_MPI::Send(sendBufferConnectivity.data(), bufferSize, MPI_CHAR, MASTER_NODE, 1, SU2_MPI::GetComm());
+    return;
+  }
+
+  /*--- Connectivity vector is written in blocks, one for each process. ---*/
+  if (nLocalElem > 0)
+    CallCGNS(cg_elements_partial_write(cgnsFileID, cgnsBase, cgnsZone, cgnsSection, firstElem, endElem,
+                                       sendBufferConnectivity.data()));
+
+  for (int i = 0; i < size; ++i) {
+    if (i == MASTER_NODE) continue;
+    /*--- In CGNS numbering starts form 1 and ranges are inclusive ---*/
+    firstElem += static_cast<cgsize_t>(distElem[i - 1]);
```

Comment:
Looks like this is the case right?
```suggestion
    firstElem = endElem + 1;
```